libass/ass_font.c

```diff
@@ -26,15 +26,165 @@
+        input_buffer[0] = 0xD800 | (code >> 10);
+        input_buffer[1] = 0xDC00 | (code & 0x3FF);
+    } else {
+        inbuf_size = 1;
+        input_buffer[0] = codepoint;
+    }
+
+    // do convert
+    char output_buffer[4];
+    BOOL conversion_fail;
+    int outbuf_size = WideCharToMultiByte(codepage, WC_NO_BEST_FIT_CHARS, input_buffer, inbuf_size,
+                                          output_buffer, sizeof(output_buffer), NULL, &conversion_fail);
+    if (outbuf_size == 0 || conversion_fail)
+        return 0;
+
+    uint32_t result = 0;
+    for (size_t i = 0; i < outbuf_size; ++i) {
+        result <<= 8;
+        result |= (uint8_t) output_buffer[i];
+    }
+
+    return result;
+}
+#elif defined(CONFIG_ICONV)
+static uint32_t convert_unicode_to_mb(FT_Encoding encoding, uint32_t codepoint)
+{
+    typedef struct { const char* names[8]; } EncodingList;
+
+    EncodingList encoding_list;
+
+    // map freetype encoding to iconv encoding
+    switch (encoding) {
+    case FT_ENCODING_MS_SJIS:
+        encoding_list = (EncodingList) {{"CP932", "SHIFT_JIS", NULL}};
+        break;
+    case FT_ENCODING_MS_GB2312:
+        encoding_list = (EncodingList) {{"CP936", "GBK", "GB18030", "GB2312", NULL}};
+        break;
+    case FT_ENCODING_MS_BIG5:
+        encoding_list = (EncodingList) {{"CP950", "BIG5", NULL}};
+        break;
+    case FT_ENCODING_MS_WANSUNG:
+        encoding_list = (EncodingList) {{"CP949", "EUC-KR", NULL}};
+        break;
+    case FT_ENCODING_MS_JOHAB:
+        encoding_list = (EncodingList) {{"CP1361", "JOHAB", NULL}};
+        break;
+    default:
+        return 0;
+    }
+
+    // open iconv context
+    const char** encoding_str = encoding_list.names;
+    iconv_t cd = (iconv_t) -1;
+    while (*encoding_str) {
+        cd = iconv_open(*encoding_str, "UTF-32LE");
+        if (cd != (iconv_t) -1) break;
+        ++encoding_str;
+    }
+    if (cd == (iconv_t) -1)
+        return 0;
+
+    char input_buffer[4];
+    char output_buffer[2]; // MS-flavour encodings only need 2 bytes
+    uint32_t result = codepoint;
+
+    // convert input codepoint to little endian uint32_t bytearray,
+    // result becomes 0 after the loop finishes
+    for (int i = 0; i < 4; ++i) {
+        input_buffer[i] = result & 0xFF;
+        result >>= 8;
+    }
+
+    // do actual convert, only reversible converts are valid, since we are converting unicode to something else
+    size_t inbuf_size = sizeof(input_buffer);
+    size_t outbuf_size = sizeof(output_buffer);
+    char* inbuf = input_buffer;
+    char* outbuf = output_buffer;
+    if (iconv(cd, &inbuf, &inbuf_size, &outbuf, &outbuf_size))
+        goto clean;
+
+    // now we have multibyte string in output_buffer
+    // assemble those bytes into uint32_t
+    size_t output_bytes = sizeof(output_buffer) - outbuf_size;
+    for (size_t i = 0; i < output_bytes; ++i) {
+        result <<= 8;
+        result |= (uint8_t) output_buffer[i];
+    }
```

Comment:
Would it be worth it to extract this little loop into its own function and reuse it in both implementations of `convert_unicode_to_mb`? Probably not significantly less code, but maybe a little less opportunity for bugs?